<a href="https://colab.research.google.com/github/2022bcs0187-sujal/lab2/blob/main/2022BCS0187_mlops_lab2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### **Name:** Chodvadiya Sujal Hasmukhbhai  
### **Roll No:** 2022BCS0187  
### **Course:** CSS 426
### **Subject:** Lab 1

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

path = "/content/drive/MyDrive/Mlops/Datasets/wine-quality/winequality-red.csv"
df = pd.read_csv(path, sep=";")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

<h2><b>Dataset Understanding</b></h2>

<h4><b>Number of Samples & Features</b></h4>

In [ ]:
print("Number of samples:", df.shape[0])
print("Number of features:", df.shape[1] - 1)

Number of samples: 1599
Number of features: 11


<h4><b>Identify Target Variable</b></h4>

In [ ]:
target = "quality"
features = df.drop(columns=[target])

<h4><b>Check Missing Values</b></h4>

In [ ]:
df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

<h4><b>Dataset Summary</b></h4>

In [ ]:
df.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1599.000000       1599.000000  1599.000000     1599.000000   
mean        8.319637          0.527821     0.270976        2.538806   
std         1.741096          0.179060     0.194801        1.409928   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.580000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1599.000000          1599.000000           1599.000000  1599.000000   
mean      0.087467            15.874922             46.467792     0.996747   
std       0.047065            10.460157             32.895324     0.001887   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996750   
75%       0.090000            21.000000             62.000000     0.997835   
max       0.611000            72.000000            289.000000     1.003690   

                pH    sulphates      alcohol      quality  
count  1599.000000  1599.000000  1599.000000  1599.000000  
mean      3.311113     0.658149    10.422983     5.636023  
std       0.154386     0.169507     1.065668     0.807569  
min       2.740000     0.330000     8.400000     3.000000  
25%       3.210000     0.550000     9.500000     5.000000  
50%       3.310000     0.620000    10.200000     6.000000  
75%       3.400000     0.730000    11.100000     6.000000  
max       4.010000     2.000000    14.900000     8.000000

<h2><b>Model Selection</b></h2>

<table>
  <tr>
  <th>Model Type</th>
  <th>Model</th>
  </tr>
  <tr>
  <td>Linear Model</td>
  <td>Linear Regression</td>
  </tr>
  <tr>
  <td>Tree-Based Model</td>
  <td>Random Forest</td>
  </tr>
</table>

<h2><b>Linear Regression Implementation</b></h2>

In [ ]:
from sklearn.metrics import mean_squared_error, r2_score

def run_experiment(model, X_train, X_test, y_train, y_test):
    model.fit(X_train, y_train)
    predictions = model.predict(X_test)
    mse = mean_squared_error(y_test, predictions)
    r2 = r2_score(y_test, predictions)
    return mse, r2

<h4><b>EXP-01: Linear Regression (All Features, No Scaling)</b></h4>

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

x = df.drop(columns=['quality'])
y = df['quality']

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

model = LinearRegression()
mse1, r2_1 = run_experiment(model, X_train, X_test, y_train, y_test)

mse1, r2_1

(0.3900251439639545, 0.40318034127962254)

<h4><b>EXP-02: Ridge Regression + Standardization + Feature Selection</h4></b>

In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge

scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

corr = df.corr()['quality'].abs()
selected_features = corr[corr > 0.2].index.drop('quality')

X_sel = df[selected_features]

X_train, X_test, y_train, y_test = train_test_split(
    X_sel, y, test_size=0.3, random_state=42
)

model = Ridge(alpha=1.0)
mse2, r2_2 = run_experiment(model, X_train, X_test, y_train, y_test)

mse2, r2_2

(0.43208823456060347, 0.31849800968816167)

<h4><b>EXP-03: Random Forest (50 Trees)</h4></b>

In [ ]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor(n_estimators=50, max_depth=10, random_state=42)

X_train, X_test, y_train, y_test = train_test_split(
    x, y, test_size=0.2, random_state=42
)

mse3, r2_3 = run_experiment(model, X_train, X_test, y_train, y_test)

mse3, r2_3

(0.3269784521147545, 0.4996549037410771)

<h4><b>Random Forest (100 Trees, Feature Subset)</h4></b>

In [ ]:
model = RandomForestRegressor(n_estimators=100, max_depth=15, random_state=42)

X_train, X_test, y_train, y_test = train_test_split(
    X_sel, y, test_size=0.2, random_state=42
)

mse4, r2_4 = run_experiment(model, X_train, X_test, y_train, y_test)

mse4, r2_4


(0.3484085827985247, 0.4668623428537646)

<table border="1" cellpadding="8" cellspacing="0" style="border-collapse: collapse; width: 100%; text-align: center;">
  <thead>
    <tr>
      <th>Experiment ID</th>
      <th>Model Type</th>
      <th>Hyperparameters</th>
      <th>Preprocessing</th>
      <th>Feature Selection Method</th>
      <th>Train/Test Split</th>
      <th>MSE</th>
      <th>R² Score</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>EXP-01</td>
      <td>Linear Regression</td>
      <td>Default</td>
      <td>None</td>
      <td>All Features</td>
      <td>80 / 20</td>
      <td>0.3900251439639545</td>
      <td>0.40318034127962254</td>
    </tr>
    <tr>
      <td>EXP-02</td>
      <td>Ridge Regression</td>
      <td>α = 1.0</td>
      <td>Standardization</td>
      <td>Correlation-based</td>
      <td>70 / 30</td>
      <td>0.43208823456060347</td>
      <td>0.31849800968816167</td>
    </tr>
    <tr>
      <td>EXP-03</td>
      <td>Random Forest</td>
      <td>50 trees, max depth = 10</td>
      <td>None</td>
      <td>All Features</td>
      <td>80 / 20</td>
      <td>0.3269784521147545</td>
      <td>0.4996549037410771</td>
    </tr>
    <tr>
      <td>EXP-04</td>
      <td>Random Forest</td>
      <td>100 trees, max depth = 15</td>
      <td>None</td>
      <td>Selected Features</td>
      <td>80 / 20</td>
      <td>0.3484085827985247</td>
      <td>0.4668623428537646</td>
    </tr>
  </tbody>
</table>


## Task 5: Analysis and Interpretation

### 1. Which experiment produced the lowest MSE?
**Experiment EXP-03** produced the lowest Mean Squared Error (MSE) with a value of **0.3269**.  
This indicates that the Random Forest model with 50 trees and depth 10 achieved the most accurate predictions among all experiments.

---

### 2. Which experiment achieved the highest R² score?
**Experiment EXP-03** achieved the highest R² score of **0.4996**, meaning it explained nearly 50% of the variance in the wine quality scores.

---

### 3. How did pre-processing affect model performance?
Pre-processing through standardization and regularization (EXP-02) resulted in **lower performance** compared to the baseline linear regression model (EXP-01).  
This suggests that scaling and regularization did not significantly benefit the linear model for this dataset, possibly due to already well-scaled features or loss of important variance.

---

### 4. Did feature selection improve or degrade performance?
Feature selection had a **mixed impact**:
- It **degraded performance** for the linear model (EXP-02), reducing both MSE and R².
- It **slightly improved performance** for the Random Forest model (EXP-04) compared to linear models, but did not surpass EXP-03 which used all features.

---

### 5. Identify the best overall model with justification.
The **Random Forest model in EXP-03** is the best overall model as it achieved the **lowest MSE** and the **highest R² score**.  
Its superior performance is attributed to its ability to capture complex, non-linear relationships in the data without requiring extensive preprocessing.


<h1><b>Why is structured experiment tracking critical for scalable and reproducible machine
learning workflows?</b><h1>

Structured experiment tracking is critical for scalable and reproducible machine learning workflows because it ensures that every model’s configuration, data split, and evaluation metric is clearly documented. As seen in this lab, multiple experiments with different models, preprocessing steps, and hyperparameters can quickly become difficult to manage without proper tracking. Systematic tracking enables fair comparison between experiments and helps identify the best-performing model reliably. It also allows experiments to be reproduced or extended in the future without ambiguity. As projects scale and involve more experiments or collaborators, structured tracking becomes essential to avoid errors and improve collaboration. Ultimately, it motivates the adoption of MLOps tools for automation and reliability.